File: daily/30 -day finish 300 coding challenge/2020-5-22. 10 questions/138. Copy List with Random Pointer.py
# This question is really good!
# solve this use DFS 
# solve this question use BFS

class Node:
    def __init__(self, x, next, random):
        self.val = int(x)
        self.next = next
        self.random = random

def class Solution:
    def __init__(self):
        self.seen = {}
    
    def copyRandomList(self, head):
        if not head: return None

        if head in self.seen:
            return self.seen[head]
        # create a new node
        node = Node(head.val, None, None)

        # create a new node
        self.seen[head] = node
        # recursion
        node.next = self.copyRandomList(head.next)
        node.random = self.copyRandomList(head.random)
        return node

# time complexity o(n)
# space complexity o(n)

        

File: JavaScript/138. Copy List With Random Pointer.js

// Clone Graph
// Clone Binary Tree With Random Pointer
// Clone N-ary Tree


var copyRandomList = function(head) {
    const map = new Map();
    function copy(node) {
        if (node == null) return null;
        if (map.get(node) != null) return map.get(node);

        const n = new Node(node.val);
        map.set(node, n);

        n.next = copy(node.next);
        n.random = copy(node.random);
        return n;
    }
    return copy(head);
}